File: meWrap/Views/StreamView/PaginatedStreamDataSource.m
//
//  PaginatedStreamDataSource.m
//  Moji
//
//  Created by Sergey Maximenko on 8/18/15.
//  Copyright (c) 2015 Ravenpod. All rights reserved.
//

#import "PaginatedStreamDataSource.h"
#import "WLStreamLoadingView.h"
#import "UIScrollView+Additions.h"
#import "WLNetwork.h"

@interface PaginatedStreamDataSource () <WLSetDelegate, WLNetworkReceiver>

@property (weak, nonatomic) WLStreamLoadingView *loadingView;

@property (nonatomic) BOOL animateLoading;

@end

@implementation PaginatedStreamDataSource

@dynamic items;

- (void)didAwake {
    [super didAwake];
    
    __weak typeof(self)weakSelf = self;
    self.loadingMetrics = [self addFooterMetrics:[[WLStreamLoadingView streamLoadingMetrics] change:^(StreamMetrics *metrics) {
        [metrics setSizeAt:^CGFloat(StreamPosition *position, StreamMetrics *metrics) {
            if (weakSelf.streamView.horizontal) {
                return weakSelf.streamView.fittingContentWidth;
            } else {
                return weakSelf.streamView.fittingContentHeight;
            }
        }];
        [metrics setHiddenAt:^BOOL(StreamPosition *position, StreamMetrics *metrics) {
            return weakSelf.items.count > 0 || weakSelf.items.completed;
        }];
        [metrics setFinalizeAppearing:^(StreamItem *item, id entry) {
            WLStreamLoadingView *loadingView = (id)item.view;
            loadingView.animating = weakSelf.animateLoading;
            weakSelf.loadingView = loadingView;
        }];
    }]];
}

- (void)setAnimateLoading:(BOOL)animateLoading {
    _animateLoading = animateLoading;
    if (self.loadingView) {
        self.loadingView.animating = animateLoading;
    }
}

- (void)setItems:(WLPaginatedSet *)items {
    [super setItems:items];
    items.delegate = self;
}

- (void)refresh {
    [self refresh:nil failure:nil];
}

- (void)refresh:(WLObjectBlock)success failure:(WLFailureBlock)failure {
    [self.items newer:success failure:failure];
}

- (BOOL)appendable {
    if (self.appendableBlock && !self.appendableBlock(self)) {
        return NO;
    }
    return self.items && !self.items.completed;
}

- (void)append:(WLObjectBlock)success failure:(WLFailureBlock)failure {
    [self.items older:success failure:failure];
}

// MARK: - WLPaginatedSetDelegate

- (void)appendItemsIfNeededWithTargetContentOffset:(CGPoint)targetContentOffset {
    StreamView *streamView = self.streamView;
    BOOL reachedRequiredOffset = NO;
    if (self.streamView.horizontal) {
        reachedRequiredOffset = (streamView.maximumContentOffset.x - targetContentOffset.x) < streamView.fittingContentWidth;
    } else {
        reachedRequiredOffset = (streamView.maximumContentOffset.y - targetContentOffset.y) < streamView.fittingContentHeight;
    }
    if (reachedRequiredOffset && [self appendable]) {
        if ([WLNetwork network].reachable) {
            self.animateLoading = YES;
            __weak typeof(self)weakSelf = self;
            [self append:nil failure:^(NSError *error) {
                weakSelf.animateLoading = NO;
                [error showIgnoringNetworkError];
            }];
        } else {
            [[WLNetwork network] addReceiver:self];
            self.animateLoading = NO;
        }
    }
}

- (void)setDidChange:(WLPaginatedSet *)group {
    [self reload];
}

- (void)streamViewDidLayout:(StreamView *)streamView {
    [super streamViewDidLayout:streamView];
    run_after_asap(^{
        [self appendItemsIfNeededWithTargetContentOffset:streamView.contentOffset];
    });
}

- (void)scrollViewWillEndDragging:(UIScrollView *)scrollView withVelocity:(CGPoint)velocity targetContentOffset:(inout CGPoint *)targetContentOffset {
    [self appendItemsIfNeededWithTargetContentOffset:*targetContentOffset];
}

// MARK: - WLNetworkReceiver

- (void)networkDidChangeReachability:(WLNetwork *)network {
    if (network.reachable) {
        [network removeReceiver:self];
        [self appendItemsIfNeededWithTargetContentOffset:self.streamView.contentOffset];
    }
}

@end

File: meWrap/Views/StreamView/PaginatedStreamDataSource.swift
//
//  PaginatedStreamDataSource.swift
//  meWrap
//
//  Created by Sergey Maximenko on 11/12/15.
//  Copyright © 2015 Ravenpod. All rights reserved.
//

import Foundation

class PaginatedStreamDataSource<T: PaginatedListProtocol>: StreamDataSource<T> {
    
    var appendableBlock: (PaginatedStreamDataSource -> Bool)?
    
    weak var spinner: UIActivityIndicatorView? {
        didSet {
            oldValue?.removeFromSuperview()
            if let spinner = spinner, let streamView = streamView {
                streamView.superview?.add(spinner, { (make) in
                    make.center.equalTo(streamView)
                })
            }
        }
    }
    
    override func didSetItems() {
        if let set = items {
            set.didChangeNotifier.subscribe(self, block: { [unowned self] (value) in
                self.reload()
                })
            set.didStartLoading.subscribe(self, block: { [unowned self] (value) in
                self.setLoading(set.count == 0 && !set.completed)
                })
            set.didFinishLoading.subscribe(self, block: { [unowned self] (value) in
                self.setLoading(false)
                })
            setLoading(set.count == 0 && !set.completed)
            
        }
        super.didSetItems()
    }
    
    func setLoading(loading: Bool) {
        guard let streamView = streamView else { return }
        let loading = loading && Network.network.reachable
        if loading != streamView.hidden {
            streamView.hidden = loading
            if loading {
                let spinner = UIActivityIndicatorView(activityIndicatorStyle: .White)
                spinner.color = Color.orange
                self.spinner = spinner
                spinner.startAnimating()
            } else {
                spinner = nil
            }
        }
    }
    
    func refresh(sender: Refresher) {
        refresh({ (_) -> Void in
            sender.setRefreshing(false, animated: true)
        }) { (error) -> Void in
            sender.setRefreshing(false, animated: true)
        }
    }
    
    func refresh() {
        refresh(nil, failure: nil)
    }
    
    func setRefreshable() {
        setRefreshableWithStyle(.White)
    }
    
    func setRefreshableWithStyle(style: RefresherStyle) {
        if let streamView = streamView {
            let refresher = Refresher(scrollView: streamView)
            refresher.style = style
            refresher.addTarget(self, action: #selector(self.refresh(_:)), forControlEvents: .ValueChanged)
        }
    }
    
    func refresh(success: ([T.PaginatedEntryType] -> ())?, failure: FailureBlock?) {
        items?.newer(success, failure: failure)
    }
    
    func append(success: ([T.PaginatedEntryType] -> ())?, failure: FailureBlock?) {
        items?.older(success, failure: failure)
    }
    
    func isAppendable() -> Bool {
        if let appendableBlock = appendableBlock where !appendableBlock(self) {
            return false
        } else {
            return !(items?.completed ?? true)
        }
    }
    
    func appendItemsIfNeededWithTargetContentOffset(targetContentOffset: CGPoint) {
        guard let streamView = streamView else { return }
        
        var reachedRequiredOffset = false
        if streamView.layout.horizontal {
            reachedRequiredOffset = (streamView.maximumContentOffset.x - targetContentOffset.x) < streamView.fittingContentWidth
        } else {
            reachedRequiredOffset = (streamView.maximumContentOffset.y - targetContentOffset.y) < streamView.fittingContentHeight
        }
        
        if reachedRequiredOffset && isAppendable() {
            if Network.network.reachable {
                append(nil, failure: { $0?.showNonNetworkError() })
            } else {
                Network.network.subscribe(self, block: { [unowned self] reachable in
                    if reachable {
                        Network.network.unsubscribe(self)
                        if let streamView = self.streamView {
                            self.appendItemsIfNeededWithTargetContentOffset(streamView.contentOffset)
                        }
                    }
                })
            }
        }
    }
    
    override func didLayout() {
        super.didLayout()
        Dispatch.mainQueue.async { () -> Void in
            self.appendItemsIfNeededWithTargetContentOffset(self.streamView!.contentOffset)
        }
    }
    
    func scrollViewWillEndDragging(scrollView: UIScrollView, withVelocity velocity: CGPoint, targetContentOffset: UnsafeMutablePointer<CGPoint>) {
        appendItemsIfNeededWithTargetContentOffset(targetContentOffset.memory)
    }
}
